### ISSUE TITLE
Hit objects may disappear in the osu!mania editor

### ISSUE THREAD

**Issue by peppy** [ppy/osu#21090]
Needs further investigation. Likely something local to object implementation on one or more skins (similar to #21089).

### Discussed in https://github.com/ppy/osu/discussions/20646

<div type='discussions-op-text'>

<sup>Originally posted by **papertek** October  9, 2022</sup>
while messing around i switched from argon to classic and went into the editor, after scrolling a bit notes started to disappear and the left key didn't stop glowing

https://user-images.githubusercontent.com/72106503/194723694-159d0a35-3935-4e81-a97f-ca00ca6b507e.mp4

</div>

**Comment by cdwcgt** [comment]
Information I now know that might help
- This problem only occurs when a note that has already been played is played again (Rewinding during replay don't occurs this error)
- Problematic notes are not available in the timeline when the timeline is zoomed up
- - `TimelineHitObjectBlurprint` just disappear from my Draw Visualiser
<details><summary>timeline zoom</summary>

before
![image](https://user-images.githubusercontent.com/34775378/201006006-d0021c35-2799-453a-bd4f-2a864659ce44.png)
after
![image](https://user-images.githubusercontent.com/34775378/201006285-81c6314a-b954-4009-95ce-ab92a4189467.png)
</details>

<details><summary>13 child but only 2 child in draw visualiser</summary>

![image](https://user-images.githubusercontent.com/34775378/201009136-93d793d3-80c8-4478-928c-7b95c2511286.png)
</details>

**Comment by cdwcgt** [comment]
https://github.com/ppy/osu/pull/21696
I encountered this situation in gameplay while developing this mod.

Objects disappear when scollvalue changes in rewind
I don't know if it's related

the conditions this failure needs:
rewind at least once (so this is replay limited) (same as editor)
Notes that have not yet appeared will not have this situation (same as editor)
**Change ScrollTime while rewind**

I tried to reproduce using the set option, but it doesn't work

This is the code I used to test
```C#
        public void Update(Playfield playfield)
        {
            currentCombo.BindValueChanged(s =>
            {
                double speed = MinScoreSpeed.Value - (MinScoreSpeed.Value - MaxScoreSpeed.Value) * Math.Log(s.NewValue + 1, MaxComboCount.Value + 1);

                scrollTime.Value = Interpolation.DampContinuously(scrollTime.Value, speed, 50, playfield.Clock.ElapsedFrameTime);
            });
        }
```

**Comment by Speykious** [comment]
If it's useful, I tried really hard to debug this issue using the visual test editor from `osu.Rulesets.Mania.Tests`, because I encounter it all the time when I'm trying to make a mania map in the lazer editor.

Unfortunately I couldn't find a single clue as to how this bug occurs after hours of code digging, but I did manage to reproduce the bug by doing the following:
- inside the Mania visual tests, go to `Editor > Editor` and press `load editor` (I also press `downwards scroll` but it's not needed)
- press space to start playing the map from the beginning inside the editor
- after about 4 seconds, _while still playing,_ seek to the beginning by clicking at the start of the `SummaryTimeline` at the bottom bar of the editor
- you'll notice a few notes and hold notes have disappeared. For me, the first 2 mania hitobjects (one hold and one note) always appear correctly, but a few ones after that do not display at all, and sometimes it seems that samples are not playing as well.

I also opened the Draw Visualizer (CTRL+F1) in hopes of maybe finding a needle in the state of the notes or of the columns. I found some weird things, but wasn't able to go from there to find were the bug originates. So when seeking to the start of the `SummaryTimeline` after a note has been played:
- the note's `State` remains `Hit` instead of going back to `Idle`
- the note's `LifetimeEnd` is a small number (in the thousands) instead of being `double.Max`, which is what it was before the object was hit the first time.

I hope this helps!

**Comment by ekrctb** [comment]
This is a regression from #20570. Nested DHOs now have precise `LifetimeStart`s in a scrolling hit object container.

1. An alive DHO can have non-alive nested DHOs. This was always possible, but this situation is more common with the PR above.
2. `DrawableHoldNote` triggers a hit result by checking results of its tail notes. <https://github.com/ppy/osu/blob/8f704249994fbc223fb5811f3cfdd0c7f9093237/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs#L255-L257>
3. A result can only be reverted while the DHO is alive because it is done in `Update`. <https://github.com/ppy/osu/blob/8f704249994fbc223fb5811f3cfdd0c7f9093237/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L581>

That means, if a hold note is rewound while tails are all dead, it incorrectly think all its tails are judged and thus a new hit result for the hold note is generated. Also, due to the hit order policy, all preceding notes are forcefully missed.
That results in notes "disappearing" because the time offset of the result is way earlier than the hit window start.

I think 3 should be fixed.

**Comment by cdwcgt** [comment]
Yes, and because of https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Mania/UI/OrderedHitPolicy.cs#L53
It will cause some notes to be mistakenly judged as miss early and disappear
I think it should be resolved after #20594 is merged?

**Comment by cdwcgt** [comment]
Separating the checking logic and adding the following logic in `DrawableHoldNote` looks fine?

```C#
protect override void Update()
{
            // ...
            if (Tail.IsHit && !Tail.IsAlive)
                Tail.CheckRevert();

            foreach (var tick in tickContainer)
            {
                if (tick.Judged && tick.IsAlive)
                    Tail.CheckRevert();
            }
}
```
